File: 프로그래머스/4/42897. 도둑질/도둑질.py
def solution(money):
    return max(dp(money, 0, len(money) - 1), dp(money, 1, len(money)))

def dp(money, start, end):
    a, b = money[start], max(money[start], money[start + 1])
    for i in range(start + 2, end):
        next = max(b, a + money[i])
        a, b = b, next
    return b
    

File: 프로그래머스/4/42897. 도둑질/도둑질.js
const dp = (money, start, end) => {
    let [first, second] = [money[start], Math.max(money[start], money[start + 1])];
    for(let i = start + 2; i <= end; i++) {
        let next = Math.max(second, first + money[i]);
        [first, second] = [second, next];
    }
    return second;
}

function solution(money) {
    return Math.max(dp(money, 0, money.length - 2), dp(money, 1, money.length - 1))
}